compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -1439,6 +1439,17 @@ class Namer { typer: Typer =>
       // println(s"owner = ${sym.owner}, decls = ${sym.owner.info.decls.show}")
       def isInlineVal = sym.isOneOf(FinalOrInline, butNot = Method | Mutable)
 
+      def isEnumValue(tp: Type) = tp.typeSymbol == defn.EnumValueClass
+
+      // Drop EnumValue parents from inferred types of enum constants
+      def dropEnumValue(tp: Type): Type = tp.dealias match
+        case tp @ AndType(tp1, tp2) =>
+          if isEnumValue(tp1) then tp2
+          else if isEnumValue(tp2) then tp1
+          else tp.derivedAndType(dropEnumValue(tp1), dropEnumValue(tp2))
```

Comment:
Same suggestion as above about dealiasing:
```suggestion
        case tpd @ AndType(tp1, tp2) =>
          if isEnumValue(tp1) then tp2
          else if isEnumValue(tp2) then tp1
          else
            val tpdw = tpd.derivedAndType(dropEnumValue(tp1), dropEnumValue(tp2))
            if tpdw ne tpd then tpdw else tp
```